### PR TITLE
Avoid NULL pointer dereference.

### LINKED ISSUE #9043: NULL pointer dereference in CSR text output code
Line 131 (in openssl 1.1.1b) of t_req.c calls X509_ATTRIBUTE_get0_type, which returns an attribute type but may return NULL. Line 132 dereferences the "type" field of the returned struct without checking for NULL. A NULL pointer dereference can be triggered given a specially crafted CSR file, causing OpenSSL to crash. Below is the PEM text for such a CSR below, which reliably causes a SIGSEGV on 1.1.1b as well as older versions of openssl by executing `openssl req -in poc.csr -text -noout`.

```
-----BEGIN CERTIFICATE REQUEST-----
MIICZjCCAU4CAQAwEjEQMA4GA1UEAwwHZm9vLmNvbTCCASIwDQYJKoZIhvcNAQEB
BQADggEPADCCAQoCggEBAMfXJrHJEUS3raD2168zPupvJcQLo0oim1Xa5qvzSEqA
NCx/fuY/b/nxbu9/bJAoPISlEjNGBgys0gdkx6r79GFfYAj2QEORRpgsibmLJmSj
99LqVOreSFVyKWlTAn9dMKwb2s3VZVZYXKphQR1Rq7C4Jm9Ip+iGooyTyG+izOC1
NRksS1hXv9dkXZHN9CUjfaKFIwW0ngkdDUfWllNxF09wJ/Uc8gAk1HCV6aGy5f4V
QwqAYlQUebhODBo+TjUcxIPZr4L4J+3Icz5ByZV2dA6c8B9j1fqS6DR0mSZoFnGE
1+0tj5BJYqV8tCGNI8iCcCUgZeHSWYUiteg8fzzZE6UCAwEAAaAPMA0GCSqGSIb3
DQEJDzEAMA0GCSqGSIb3DQEBCwUAA4IBAQB2XvvV+IBn64FS8vJExj0Yxn0ENgpX
O95jQ5yaosXlGoGxeNu5mMykO+Wko20M/HBnzCZs8Py36peCBrBtZRmpflYAksan
PH8oiEAYkn6lpeNVQJ6Byf3IfdBnkbWbO9JhY30Lb8w14z9+cJ9F6PM+ViiL3cXO
GPzmNqWyuEMQ5TG0AZK5vInhGsO/Gi4PclGFpr4AkFy+POYuxgnatzelkvPHSruM
wTUDZae3AcscywunI1uodCpbNRf5P+FR14zIC0DA/sYiYwl3+TkhkUId5Ft/wbk0
06VcPZ9qdk1cZkVTtI5fSL/oIUhWSXo6RIcZcuiwP48KE51dqmKBLZa8
-----END CERTIFICATE REQUEST-----
```

### PR DESCRIPTION
**Pull request by beldmit**
Fixes #9043.
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/x509/t_req.c [hunk 1]
```diff
@@ -129,8 +129,10 @@ int X509_REQ_print_ex(BIO *bp, X509_REQ *x, unsigned long nmflags,
...
  get_next:
                     at = X509_ATTRIBUTE_get0_type(a, ii);
-                    type = at->type;
-                    bs = at->value.asn1_string;
+                    if (at != NULL) {
+                        type = at->type;
+                        bs = at->value.asn1_string;
+                    }
                 }
                 for (j = 25 - j; j > 0; j--)
```
Comment (slontis) [inline crypto/x509/t_req.c]: potentially bs could be NULL and then dereferenced bs->length further down will now fall over.

Comment (slontis) [reply crypto/x509/t_req.c]: If in the first loop at != NULL then the type and bs will be set.
On the second loop if att is NULL then bs and type will still be set to the previous values.. Is this what you want to happen?

**Comment by kroeckx** [comment]
You should put the "Fixes #9043" on a line of itself.

**Comment by kroeckx** [comment]
Oh, github seems to understand this too.

File: crypto/x509/t_req.c (hunk 1 above)
Comment (kroeckx) [reply crypto/x509/t_req.c]: I doubt it. Either NULL is invalid here and we should return (or print) some error, or it's valid and print that.

Comment (beldmit) [reply crypto/x509/t_req.c]: I think that setting type to 0 (causing a diagnostic print) will be enough here.

Comment (beldmit) [reply crypto/x509/t_req.c]: > The get_next: construct looks a bit funny in here (looks like a for loop :))

It looks weird, but I'm not ready to refactor this place without a relevant example.

Comment (kroeckx) [reply crypto/x509/t_req.c]: The problem is not that at == NULL, the problem is that count == 0.

Comment (kroeckx) [reply crypto/x509/t_req.c]: And that count == 0 seems to be invalid.

**Comment by kroeckx** [comment]
So as I understand it, there is a set of attributes. You get the amount of attributes with X509_REQ_get_attr_count(). In that set, you have a sequence, with a type, and a set of values. The amount of values is what X509_ATTRIBUTE_count() returns, and we're iterating over that with a goto get_next; Per rfc2986 it should contain at least 1 value, but the example file contains 0 values.

**Comment by beldmit** [comment]
Checking count() causes invalid output and does not cause any diagnostic print about erroneous req. So I will add the check for non-zero X509_ATTRIBUTE_count() and a result of X509_REQ_print_ex/X509_print_ex in apps/req.c

File: crypto/x509/t_req.c [hunk 2]
```diff
@@ -129,8 +129,11 @@ int X509_REQ_print_ex(BIO *bp, X509_REQ *x, unsigned long nmflags,
                     count = X509_ATTRIBUTE_count(a);
  get_next:
                     at = X509_ATTRIBUTE_get0_type(a, ii);
-                    type = at->type;
-                    bs = at->value.asn1_string;
+                    if (at != NULL) {
+                        type = at->type;
+                        bs = at->value.asn1_string;
+                    } else
+                        type = 0;
```
Comment (slontis) [inline crypto/x509/t_req.c]: NIT: should be 
else {
    type = 0;
}

**Comment by levitte** [comment]
Ping.  @beldmit ?

**Comment by beldmit** [comment]
Will rewrite it and return an error in the case similar to this.

**Comment by beldmit** [comment]
Ping?

**Comment by levitte** [comment]
You gonna merge this, @paulidale?

**Comment by paulidale** [comment]
Merged to 1.1.1 and master.

1.1.0 wasn't clean, see #9322.

**Pull request closed without merging**